### ISSUE TITLE
Webhook integration fails with "response code is 400"

### ISSUE THREAD

**Issue by AlexBarth13** [goharbor/harbor#16632]
**Expected behavior and actual behavior:**
Hi all,

Currently we are facing an issue with the webhook integration in Harbor. We want to send outbound messages to our internal Mattermost application. I have set up a Mattermost Incoming Webhook for a private and public channel. For both channels we will see the following error messages in the logs:

`Job 'WEBHOOK:0490888e0d6e06a2ba632b98' exit with error: run error: webhook job(target: https://URL/hooks/xxxx) response code is 400`

The DEBUG log level is enabled.

Here is a screenshot, of the webhook I set up in Harbor:

![image](https://user-images.githubusercontent.com/99802718/161249789-80534546-5df8-4324-95b6-23fad726f3e9.png)

**Steps to reproduce the problem:**
- Set up Incoming webhook in Mattermost
- Create webhook in a Harbor project with type `http`
- Add the Mattermost Webhook URL
- Select `Verify Remote Certificate`
- Pull or push an image from the project

**Versions:**
- harbor version: v2.4.0
- Mattermost: latest version

<br ><br >
Are you aware of this or do you know how to fix this issue? Thank you in advance!
<br ><br >

Alexander Barth (alexander.barth@mercedes-benz.com) on behalf of Daimler TSS, [Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Comment by wy65701436** [comment]
please provide the jobservice log as well, and can you ping the webhook success inside the jobservice container?

**Comment by AlexBarth13** [comment]
Hi @wy65701436 Thanks for your reply!
From within the jobservice container I'm able to reach our Mattermost endpoint in port 443 with a curl command.
Please find below the log entries from the jobservice.log file:

```
2022-04-04 16:52:31	
Apr  4 16:52:31 <IP> jobservice[15133]: 2022-04-04T14:52:31Z [ERROR] [/jobservice/runner/redis.go:122]: Job 'WEBHOOK:0c788d7ad979f53b52c4f8cb' exit with error: run error: webhook job(target: https://<MATTERMOST_URL>/hooks/xxxx) response code is 400

2022-04-04 16:52:31
Apr  4 16:52:31 <IP> jobservice[15133]: 2022-04-04T14:52:31Z [ERROR] [/jobservice/job/impl/notification/webhook_job.go:62]: webhook job(target: https://<MATTERMOST_URL>/hooks/xxxx) response code is 400

2022-04-04 16:52:31	
Apr  4 16:52:31 <IP> jobservice[15133]: 2022-04-04T14:52:31Z [DEBUG] [/jobservice/hook/hook_agent.go:118]: Hook event is successfully sent: status change: job=0c788d7ad979f53b52c4f8cb, status=Running, revision=1649083951->http://core:8080/service/notifications/jobs/webhook/442

2022-04-04 16:52:31	
Apr  4 16:52:31 <IP> jobservice[15133]: 2022-04-04T14:52:31Z [INFO] [/jobservice/worker/cworker/c_worker.go:76]: Job incoming: {"name":"WEBHOOK","id":"0c788d7ad979f53b52c4f8cb","t":1649083950,"args":null,"fails":1,"err":"run error: webhook job(target: https://<MATTERMOST_URL>/hooks/xxxx) response code is 400","failed_at":1649083914}

Apr  4 16:51:54 <IP> jobservice[3768]: 2022-04-04T14:51:54Z [DEBUG] [/jobservice/api/handler.go:278]: Serve http request 'POST /api/v1/jobs': 202 {"job":{"id":"0c788d7ad979f53b52c4f8cb","status":"Pending","name":"WEBHOOK","kind":"Generic","unique":false,"ref_link":"/api/v1/jobs/0c788d7ad979f53b52c4f8cb","enqueue_time":1649083914,"update_time":1649083914,"web_hook_url":"http://core:8080/service/notifications/jobs/webhook/442","parameters":{"address":"https://<MATTERMOST_URL>/hooks/xxxx","auth_header":"","payload":"{\"type\":\"PULL_ARTIFACT\",\"occur_at\":1649083914,\"operator\":\"<USER>\",\"event_data\":{\"resources\":[{\"digest\":\"sha256:xxxx\",\"tag\":\"sha256:xxxx\",\"resource_url\":\"<HARBOR_URL>/<USER>/<IMAGE>@sha256:xxxx\"}],\"repository\":{\"date_created\":1648570422,\"name\":\"<IMAGE>\",\"namespace\":\"<USER>\",\"repo_full_name\":\"<USER>/<IMAGE>\",\"repo_type\":\"private\"}}}","skip_cert_verify":false}}}
```

Thanks!
<br ><br >

Alexander Barth (alexander.barth@mercedes-benz.com) on behalf of Daimler TSS, [Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Comment by wy65701436** [comment]
@AlexBarth13 can you try to send a post request to the webhook inside the jobservice container? Let's see the response.

**Comment by AlexBarth13** [comment]
@wy65701436 
Sending out a post request is working fine and the message will reach my private channel in Mattermost.
I have tested it with this command:

```
curl -i -X POST -H 'Content-Type: application/json' -d '{"text": "Hello, this is some text\nThis is more text. :tada:"}'
<MATTERMOST_WEBHOOK_URL>
```

This was the output of it:
```
Thu Apr  7 06:02:01 UTC 2022
HTTP/1.1 200 OK
Server: nginx
Date: Thu, 07 Apr 2022 05:57:40 GMT
Content-Type: text/plain
Content-Length: 2
Connection: keep-alive
X-Request-Id: xxxxx
X-Version-Id: xxxxx
Strict-Transport-Security: max-age=63072000; includeSubdomains
X-Content-Type-Options: nosniff
Strict-Transport-Security: max-age=63072000; includeSubdomains
X-Content-Type-Options: nosniff

ok
```

And the message reach my channel:

![image](https://user-images.githubusercontent.com/99802718/162131081-9be32b5e-2875-4724-9780-2e498ffd4d93.png)


Thanks!
<br ><br >

Alexander Barth (alexander.barth@mercedes-benz.com) on behalf of Daimler TSS, [Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Comment by AlexBarth13** [comment]
Below you can find a detailed anaylsis of the problem and why we can see the `400` error code.

**TL;DR:**
- Mattermost expects another format of the message and hence will send a `400` error code
- Mattermost expects a "text" parameter in the message body
- Harbor don't send a "text" parameter and therefore, Mattermost reject the message

**Details:**
### Harbor message

The message body, which will be generated from Harbor can be found unter `parameters`:

```
May 16 16:31:55 192.168.160.1 jobservice[1732]: 2022-05-16T14:31:55Z [DEBUG] [/jobservice/api/handler.go:279]: Serve http request 'POST /api/v1/jobs': 202 {
	"job":{
		"id":"060ad517a8207ff3db7556a9",
		"status":"Pending",
		"name":"WEBHOOK",
		"kind":"Generic"
		,"unique":false,
		"ref_link":"/api/v1/jobs/060ad517a8207ff3db7556a9",
		"enqueue_time":1652711515,
		"update_time":1652711515,
		"web_hook_url":"http://core:8080/service/notifications/jobs/webhook/90",
		"parameters":{
			"address":"https://<MATTERMOST_URL>/hooks/<TOKEN>",
			"auth_header":"",
			"payload":"
			{
				\"type\":\"DELETE_ARTIFACT\",
				\"occur_at\":1652711515,
				\"operator\":\"USERID\",
				\"event_data\":
				{
					\"resources\":
					[
						{
							\"digest\":\"sha256:61face6bf030edce7ef6d7dd66fe452298d6f5f7ce032afdd01683ef02b2b841\",
							\"tag\":\"latest\",
							\"resource_url\":\"<REGISTRY>/<PROJECT>/nginx:latest\"
						}
					],
						\"repository\":
						{
							\"name\":\"nginx\",
							\"namespace\":\"<PROJECT>\",
							\"repo_full_name\":\"<PROJECT>/nginx\",
							\"repo_type\":\"public\"
						}
				}
			}",
			"skip_cert_verify":false
		}
	}
}
```

## curl testing
If we put everything in between `parameters` into an curl command the following error message will be seen.

**Command:**
```
curl -i -X POST -H 'Content-Type: application/json' -d '{\"type\":\"DELETE_ARTIFACT\",\"occur_at\":1652711515,\"operator\":\"<PROJECT>\",\"event_data\":{\"resources\":[{\"digest\":\"sha256:61face6bf030edce7ef6d7dd66fe452298d6f5f7ce032afdd01683ef02b2b841\",\"tag\":\"latest\",\"resource_url\":\"<REGISTRY>/<PROJECT>/nginx:latest\"}],\"repository\":{\"name\":\"nginx\",\"namespace\":\"<PROJECT>\",\"repo_full_name\":\"<PROJECT>/nginx\",\"repo_type\":\"public\"}}}' https://<MATTERMOST_URL>/hooks/<TOKEN>
```

**Error:**
```
Mo 16. Mai 16:41:23 CEST 2022
HTTP/1.1 400 Bad Request
Server: nginx
Date: Mon, 16 May 2022 14:35:25 GMT
Content-Type: application/json
Content-Length: 172
Connection: keep-alive
X-Request-Id: xxx178xikfd8dxpg1rfntd5cqy
X-Version-Id: 6.6.1.6.6.1.3fb03409381b1e9a528beeca6f24746a.true
Strict-Transport-Security: max-age=63072000; includeSubdomains
Strict-Transport-Security: max-age=63072000; includeSubdomains

{"id":"model.incoming_hook.parse_data.app_error","message":"Unable to parse incoming data.","detailed_error":"","request_id":"xxx178xikfd8dxpg1rfntd5cqy","status_code":400}
```

If we remove " \ " in the body than this will happen:

**Command:**
```
curl -i -X POST --data-urlencode 'payload={"type":"DELETE_ARTIFACT","occur_at":1652711515,"operator":"<PROJECT>"}' https://<MATTERMOST_URL>/hooks/<TOKEN>
```

**Error:**
```
HTTP/1.1 400 Bad Request
Server: nginx
Date: Mon, 16 May 2022 14:56:17 GMT
Content-Type: application/json
Content-Length: 155
Connection: keep-alive
X-Request-Id: hdc9mnm4xf8g9engeoy4i7nike
X-Version-Id: 6.6.1.6.6.1.3fb03409381b1e9a528beeca6f24746a.true
Strict-Transport-Security: max-age=63072000; includeSubdomains
Strict-Transport-Security: max-age=63072000; includeSubdomains

{"id":"web.incoming_webhook.text.app_error","message":"No text specified.","detailed_error":"","request_id":"hdc9mnm4xf8g9engeoy4i7nike","status_code":400}
```
If we change the message and add a `"text"` parameter into it, the message will be send to Mattermost:

**Command:**
```
curl -i -X POST --data-urlencode 'payload={"text": "type: DELETE_ARTIFACT,\noccur_at: 1652711515,\noperator: <PROJECT>"}' https://<MATTERMOST_URL>/hooks/<TOKEN>
```

**Error:**
```
HTTP/1.1 200 OK
Server: nginx
Date: Mon, 16 May 2022 14:59:37 GMT
Content-Type: text/plain
Content-Length: 2
Connection: keep-alive
X-Request-Id: p97fizysgpf5fypjxunp418m6r
X-Version-Id: 6.6.1.6.6.1.3fb03409381b1e9a528beeca6f24746a.true
Strict-Transport-Security: max-age=63072000; includeSubdomains
X-Content-Type-Options: nosniff
Strict-Transport-Security: max-age=63072000; includeSubdomains
X-Content-Type-Options: nosniff

ok
```

Does anyone can help here or know how to fix it?

Thanks a lot!
<br ><br >
Alexander Barth (alexander.barth@mercedes-benz.com) on behalf of Mercedes-Benz Tech Innovation GmbH, [Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Comment by AlexBarth13** [comment]
The issue still exists and needs to be fixed.

**Comment by JonnyBDev** [comment]
+1 Need this implementation for our Mattermost service too.

**Comment by pandel** [comment]
+1 Will this be fixed in a reasonable amount of time?

I am using Harbor version 2.6.1. Today I tried to use "slack" as the notify type, as given in your own documentation (Mattermost is mostly Slack compatible), but that also resulted in error 400. The mattermost server told me that no text was given.

So the problem exists not only with type "http" but also with type "slack". This should be fixed.

**Comment by pandel** [comment]
I thought about this problem a little longer and until this is fixed somehow officially, I tried to create some kind of bridge between Harbor and Mattermost and came up with the script attached. This is more like a skeleton atm, but I think one can get the idea.

You  start the skript like this: `python3 echo_server.py <port to listen to> <Mattermost Server URL> <api_key>`
and you have to set your Harbor Endpoint URL to `http://<machine the script is running on>:<port to listen to>`

It will listen for incoming messages, reformat them and then send them to you Mattermost Server. Yes, I know, no authentication, no SSL, only pull and replication events with more details atm, no nothing, but it is an idea.

It is mostly based on a gist I do not find anymore and the fabulous https://github.com/numberly/matterhook library, so you need to install that first: `pip install matterhook`

Perhaps someone can build something more sophisticated based on this...

<details><summary>Python message bridge skeleton:</summary>

```
#!/usr/bin/env python3

from http.server import BaseHTTPRequestHandler, HTTPServer
from matterhook import Webhook
from datetime import datetime
import json
import logging

class MatterBridge(BaseHTTPRequestHandler):
    matterserver = ''
    api_key = ''

    def _set_response(self):
        self.send_response(200)
        self.send_header('Content-type', 'text/html')
        self.end_headers()

    def do_GET(self):
        # only for informational purpose
        logging.info("GET request,\nPath: %s\nHeaders:\n%s\n", str(self.path), str(self.headers))
        self._set_response()

    def do_POST(self):
        mwh = Webhook(self.matterserver,self.api_key)
        content_length = int(self.headers['Content-Length']) # <--- Gets the size of data
        post_data = self.rfile.read(content_length) # <--- Gets the data itself

        # Activate logging for every request
        #logging.info("POST request,\nPath: %s\nHeaders:\n%s\n\nBody:\n%s\n",
        #        str(self.path), str(self.headers), post_data.decode('utf-8'))

        # print pure data
        # print(json.dumps(data,indent=4))

        data=json.loads(post_data.decode('utf-8'),strict=False)

        try:
            message=''
            if data['type'] == 'PUSH_ARTIFACT':
                message='Image pushed'
            elif data['type'] == 'PULL_ARTIFACT':
                operator=data['operator']
                occur_at=datetime.strftime(datetime.fromtimestamp(data['occur_at']),"%c")
                image=data['event_data']['repository']['repo_full_name']
                message='Pulled image:\n'
                message+=image+'\n'
                message+='Operator: ' + operator + '\n'
                message+='Event timestamp: ' + occur_at
            elif data['type'] == 'DELETE_ARTIFACT':
                message='Artifact deleted'
            elif data['type'] == 'UPLOAD_CHART':
                message='Chart uploaded'
            elif data['type'] == 'DOWNLOAD_CHART':
                message='Chart downloaded'
            elif data['type'] == 'DELETE_CHART':
                message='Chart deleted'
            elif data['type'] == 'SCANNING_COMPLETED':
                message='Scanning completed'
            elif data['type'] == 'SCANNING_FAILED':
                message='Scanning failed'
            elif data['type'] == 'QUOTA_EXCEED':
                message='Quota exceeded'
            elif data['type'] == 'QUOTA_WARNING':
                message='Quota warning'
            elif data['type'] == 'REPLICATION':
                status=data['event_data']['replication']['job_status']
                if status == 'Success':
                    operator=data['operator']
                    occur_at=datetime.strftime(datetime.fromtimestamp(data['occur_at']),"%c")
                    namespace=data['event_data']['replication']['dest_resource']['namespace']
                    message='Image replicated:\n'
                    message+='Namespace: ' + namespace +'\n'
                    if 'successful_artifact' in data['event_data']['replication']:
                        for part in data['event_data']['replication']['successful_artifact']:
                            message+='Artifact: ' + part['name_tag'] + '\n'
                    message+='Operator: ' + operator + '\n'
                    message+='Status: ' + status + '\n'
                    message+='Event timestamp: ' + occur_at

            if message != '':
                mwh.send(message)

        except:
            # show requests with problems
            logging.info("POST request,\nPath: %s\nHeaders:\n%s\n\nBody:\n%s\n",
                    str(self.path), str(self.headers), post_data.decode('utf-8'))
            # print raw body data
            # print(json.dumps(data, indent=4))

        self._set_response()

def run(server_class=HTTPServer, handler_class=MatterBridge, port=8065, mm='', key=''):
    logging.basicConfig(level=logging.INFO)

    MatterBridge.matterserver = mm
    MatterBridge.api_key = key
    
    server_address = ('', port)
    httpd = server_class(server_address, handler_class)

    logging.info('Starting httpd...\n')
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        pass
    httpd.server_close()
    logging.info('Stopping httpd...\n')

if __name__ == '__main__':
    from sys import argv

    if len(argv) < 4:
        print("Harbor to Mattermost Webhook Bridge")
        print("Usage:")
        print("./echo_server.py <port> '<Mattermost URL>' '<API KEY>'")
        print()
        print("Port, Mattermost Server URL, API KEY missing")
    else:
        run(port=int(argv[1]), mm=str(argv[2]), key=str(argv[3]))
```

 </details>

**Comment by ntarocco** [comment]
I am also interested in this. I can create a PR for Mattermost and Discord, if you are willing to merge it and integrate it in one of the next releases.

**Comment by wy65701436** [comment]
I personally don't think it is a bug of harbor that needs to be fixed since Mattermost requires a "text" parameter in the message body which is not by design in the harbor side. Maybe you can set up a service to wrap or customize the webhook.

@AlexBarth13 should we close it?

**Comment by AlexBarth13** [comment]
Hi @wy65701436 
I think this is not a bug of Harbor, but I (and others as you can above) would like to have this possibility to send messages to Mattermost. I could think about something in addition to the existing notify types `http` and ` slack` in the dropdown menu (e.g. `mattermost`).
In our organization thousands of people are using Mattermost and we as a platform team have a lot of customers who would like to have this feature/possibility.
Therefore I would like to keep this issue open and hope, this could be something for the v2.7 release.

<br ><br >
Alexander Barth (alexander.barth@mercedes-benz.com) on behalf of Mercedes-Benz Tech Innovation GmbH, [Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Comment by JonnyBDev** [comment]
I completely agree with @AlexBarth13 . This is not a bug. But if someone like @ntarocco would crated a PR for Discord and Mattermost, this would benefit Harbor. A lot of companies, mine as well, are using Mattermost because of compliance and on premise hosting. It would be great to see a software like Harbor supporting the most common providers.

**Comment by AlexBarth13** [comment]
Hi @wy65701436
This feature request is still valid. Is it maybe possible to implement it in 2.8.0?

Thanks!
<br ><br >
Alexander Barth (alexander.barth@mercedes-benz.com) on behalf of Mercedes-Benz Tech Innovation GmbH, [Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Comment by wy65701436** [comment]
@AlexBarth13  we're considering to do some refactor work for webhook in v2.8.0 in order to support CloudEvent, customize payload could be a candidate item. cc @chlins

**Comment by rohitghegde** [comment]
This seems to be the isssue for many, our organization also has the similar setup and we are trying to configure notifications about vulneribility scanning into our mattermost channel. 

When can we expect the v2.8.0 release with the fix  or new feature to choose the harbor payload type?

**Comment by JonnyBDev** [comment]
The v2.8 release featured the implementation of "Cloudevents" (https://cloudevents.io/) payload. Seems like the direction is "tell Mattermost to adjust their payload format to a proper one" instead of adding the actual Mattermost payload definition.

**Comment by iva-5-middleware** [comment]
Hi @wy65701436,

I don't really understand this thread and the point of view, that Mattermost has to change their payload syntax. If there was the ability to self-define the payload (so type "custom webhook", instead of "slack", "mattermost" or others), there would be no problem at all. And even other software with different payload syntaxes would profit a lot from this ability.
I think that's the easiest way to implement this and it shouldn't need much effort to realize.

e.g. Jira Automation does support custom webhook payload like this:
![jira_automation_custom_payload](https://github.com/user-attachments/assets/d633f2ac-6b49-46a2-9058-3c846683eb1e)

It's that easy.

**Comment by worksserhat** [comment]
+1 Need this implementation.

**Comment by strassheim** [comment]
+1 it would be nice to have a generic form to fill out data myself - as described by [iva-5-middleware](https://github.com/iva-5-middleware)

please + thxia
